### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [8]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data_pd = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [9]:
#Calculate total number of players and create dataframe
total_players = len(purchase_data_pd["SN"].unique())
total_players_df = pd.DataFrame({"Total Players": [total_players]})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [20]:
#Calculate number of unique items, average price, purchase count, and total revenue
item_ID = len(purchase_data_pd["Item ID"].unique())
average_price = purchase_data_pd["Price"].mean()
purchase_count = purchase_data_pd["Purchase ID"].count()
total_revenue = purchase_data_pd["Price"].sum()

#Create dataframe for number of unique items, average price, number of purchases, and total revenue
summary_df = pd.DataFrame({"Number of Unique Items": [item_ID],
                              "Average Price": [average_price],
                              "Number of Purchases": [purchase_count],
                              "Total Revenue": [total_revenue]})

#Format Average Price and Total Revenue in dataframe
summary_df["Average Price"] = summary_df["Average Price"].map("${:,.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:,.2f}".format)
summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [26]:
#Calculate percentage and count of all genders and create dataframe

demographics = purchase_data_pd[["SN","Age","Gender"]]
unique_num_bag = demographics.drop_duplicates()

gender_total = unique_num_bag["Gender"].value_counts()

percent = purchase_data_pd["Gender"].value_counts(normalize=True).mul(100).round(1)

gender_demog_df = pd.DataFrame({"Count": gender_total, "Percent": percent})
gender_demog_df["Percent"] = gender_demog_df["Percent"].astype(str) + "%"
gender_demog_df

Count Percent
Male                     484   83.6%
Female                    81   14.5%
Other / Non-Disclosed     11    1.9%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [18]:
#Calculate average purchase price, total purchase price, and average total purchase per person by gender
pu_ct = purchase_data_pd.groupby("Gender")["Price"].count()

avg_pp_bg = purchase_data_pd.groupby("Gender")["Price"].mean().round(2)
total_pp_bg = purchase_data_pd.groupby("Gender")["Price"].sum()

avg_tot_pp_bg = (total_pp_bg/gender_total).round(2)

#Create dataframe for purchase count, average purchase price, total purchase value, and average total purchase per person
purchasing_anal_df = pd.DataFrame({"Purchase Count": pu_ct, "Average Purchase Price": avg_pp_bg, 
                                   "Total Purchase Value": total_pp_bg, "Avg Total Purchase per Person": avg_tot_pp_bg})

#Format Average Purchase Price, Total Purchase Value, and Avg Total Purchase per Person in dataframe
purchasing_anal_df["Average Purchase Price"] = purchasing_anal_df["Average Purchase Price"].map("${:,.2f}".format)
purchasing_anal_df["Total Purchase Value"] = purchasing_anal_df["Total Purchase Value"].map("${:,.2f}".format)
purchasing_anal_df["Avg Total Purchase per Person"] = purchasing_anal_df["Avg Total Purchase per Person"].map("${:,.2f}".format)
purchasing_anal_df





Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [31]:
#Establish bins for ages and group labels for ages
bins = [0,9,14,19,24,29,34,39,50]
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39","40+"]

#Categorize players by group labels and calculate percentages by age group

purchase_data_pd["Age Group"] = pd.cut(purchase_data_pd["Age"], bins, labels=group_labels)

#Determine number of unique players in age group
age_demographics = purchase_data_pd[["SN","Age Group"]]

unique_demo = age_demographics.drop_duplicates()

age_demo_grp = unique_demo["Age Group"].value_counts()

#Calculate sum of unique players in all age groups
age_demo_total = age_demo_grp.sum()

percent_age_grp = ((age_demo_grp/age_demo_total)*100).round(1)

#Create dataframe for total count and percentage of players by age group
age_grp_df = pd.DataFrame({"Total Count": age_demo_grp, "Percentage of Players": percent_age_grp})
age_grp_df["Percentage of Players"] = age_grp_df["Percentage of Players"].astype(str) + "%"
age_grp_df


Total Count Percentage of Players
20-24          258                 44.8%
15-19          107                 18.6%
25-29           77                 13.4%
30-34           52                  9.0%
35-39           31                  5.4%
10-14           22                  3.8%
<10             17                  3.0%
40+             12                  2.1%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [17]:
#Determine unique count of players in each age group

age_grp_pur_ct = purchase_data_pd[["SN","Age Group"]]
unique_ply_age_grp = age_grp_pur_ct.drop_duplicates()

age_grp_tot_pur_ct = unique_ply_age_grp["Age Group"].value_counts()

#Calculate average purchase price and total purchase value
avg_pur_prc = purchase_data_pd.groupby("Age Group")["Price"].mean().round(2)
total_pur_val = purchase_data_pd.groupby("Age Group")["Price"].sum().round(2)


#Determine unique count of players to calculate average total purchase per person
pur_per = purchase_data_pd[["SN","Age Group"]]
unique_per = pur_per.drop_duplicates()

pur_per_age_grp = unique_per["Age Group"].value_counts()

avg_total_pur_pp = (total_pur_val/pur_per_age_grp).round(2)

#Create dataframe for purchase count, average purchase price, total purchase value, and average total purchase per person
purchase_age_df = pd.DataFrame({"Purchase Count": age_grp_tot_pur_ct, "Average Purchase Price": avg_pur_prc, 
                                   "Total Purchase Value": total_pur_val, "Avg Total Purchase per Person": avg_total_pur_pp})    

#Format Average Purchase Price, Total Purchase Value, and Avg Total Purchase per Person in dataframe
purchase_age_df["Average Purchase Price"] = purchase_age_df["Average Purchase Price"].map("${:,.2f}".format)
purchase_age_df["Total Purchase Value"] = purchase_age_df["Total Purchase Value"].map("${:,.2f}".format)
purchase_age_df["Avg Total Purchase per Person"] = purchase_age_df["Avg Total Purchase per Person"].map("${:,.2f}".format)
purchase_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                17                  $3.35               $77.13   
10-14              22                  $2.96               $82.78   
15-19             107                  $3.04              $412.89   
20-24             258                  $3.05            $1,114.06   
25-29              77                  $2.90              $293.00   
30-34              52                  $2.93              $214.00   
35-39              31                  $3.60              $147.67   
40+                12                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [93]:
#Group by SN and calculate purchase count, average purchase price, and total purchase price by top spenders
pur_ct_tp = purchase_data_pd.groupby("SN")["Price"].count()
avg_pp_tp = purchase_data_pd.groupby("SN")["Price"].mean().round(2)
total_pp_tp = purchase_data_pd.groupby("SN")["Price"].sum().round(2)

#Create dataframe containing purchase count, average purchase price, and total purchase price and sort by total purchase value in descending order
top_spender_df = pd.DataFrame({"Purchase Count": pur_ct_tp, "Average Purchase Price": avg_pp_tp, "Total Purchase Value": total_pp_tp})


#Format Average Purchase Price and Total Purchase Value in dataframe
#top_spender_df["Average Purchase Price"] = top_spender_df["Average Purchase Price"].map("${:,.2f}".format)
#top_spender_df["Total Purchase Value"] = top_spender_df["Total Purchase Value"].map("${:,.2f}".format)

top_spender_df.head()
top_spender_df.sort_values(["Total Purchase Value"], ascending=False)



Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                    3.79                 18.96
Idastidru52               4                    3.86                 15.45
Chamjask73                3                    4.61                 13.83
Iral74                    4                    3.40                 13.62
Iskadarya95               3                    4.37                 13.10
...                     ...                     ...                   ...
Ililsasya43               1                    1.02                  1.02
Irilis75                  1                    1.02                  1.02
Aidai61                   1                    1.01                  1.01
Chanirra79                1                    1.01                  1.01
Alo38                     1                    1.00                  1.00

[576 rows x 3 columns]

In [89]:
#Group by SN and calculate purchase count, average purchase price, and total purchase price by top spenders
pur_ct_tp = purchase_data_pd.groupby("SN")["Price"].count()
avg_pp_tp = purchase_data_pd.groupby("SN")["Price"].mean().round(2)
total_pp_tp = purchase_data_pd.groupby("SN")["Price"].sum().round(2)

#Create dataframe containing purchase count, average purchase price, and total purchase price and sort by total purchase value in descending order
top_spender_df = pd.DataFrame({"Purchase Count": pur_ct_tp, "Average Purchase Price": avg_pp_tp, "Total Purchase Value": total_pp_tp})
#top_spender_df["Average Purchase Price"] = top_spender_df["Average Purchase Price"].map("${:,.2f}".format)
#top_spender_df["Total Purchase Value"] = top_spender_df["Total Purchase Value"].map("${:,.2f}".format)
top_spender_df.head()
top_spender_df.sort_values(["Total Purchase Value"], ascending=False)




Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                    3.79                 18.96
Idastidru52               4                    3.86                 15.45
Chamjask73                3                    4.61                 13.83
Iral74                    4                    3.40                 13.62
Iskadarya95               3                    4.37                 13.10
...                     ...                     ...                   ...
Ililsasya43               1                    1.02                  1.02
Irilis75                  1                    1.02                  1.02
Aidai61                   1                    1.01                  1.01
Chanirra79                1                    1.01                  1.01
Alo38                     1                    1.00                  1.00

[576 rows x 3 columns]

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [35]:
#Group items by Item ID and Item name--calculate purchase count, item price and total purchase value
pur_ct_mp = purchase_data_pd.groupby(["Item ID","Item Name"])["Price"].count()
item_price = purchase_data_pd.groupby(["Item ID", "Item Name"])["Price"].mean()
total_pp_mp = purchase_data_pd.groupby(["Item ID","Item Name"])["Price"].sum().round(2)

#Create dataframe for Purchase Count, Item Price, and Total Purchase Value
#Format Item Price and Total Purchase Value--Sort by Purchase Count in descending order
new_mos_pop_df = pd.DataFrame({"Purchase Count": pur_ct_mp, "Item Price": item_price, "Total Purchase Value": total_pp_mp})
new_mos_pop_df["Item Price"] = new_mos_pop_df["Item Price"].map("${:,.2f}".format)
new_mos_pop_df["Total Purchase Value"] = new_mos_pop_df["Total Purchase Value"].map("${:,.2f}".format)
new_mos_pop_df.head()
new_mos_pop_df.sort_values(["Purchase Count"], ascending=False)


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   
...                                                              ...   
104     Gladiator's Glaive                                         1   
23      Crucifer                                                   1   
180     Stormcaller                                                1   
91      Celeste                                                    1   
134     Undead Crusader                                            1   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   
...                                                         ...   
104     Gladiator's Glaive                                $1.93   
23      Crucifer                                          $1.99   
180     Stormcaller                                       $3.36   
91      Celeste                                           $4.17   
134     Undead Crusader                                   $4.50   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16  
...                                                                   ...  
104     Gladiator's Glaive                                          $1.93  
23      Crucifer                                                    $1.99  
180     Stormcaller                                                 $3.36  
91      Celeste                                                     $4.17  
134     Undead Crusader                                             $4.50  

[183 rows x 3 columns]

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [92]:
#Create dataframe for Purchase Count, Item Price, and Total Purchase Value--Sort by Total Purchase Value in descending order
new_mos_prof_df = pd.DataFrame({"Purchase Count": pur_ct_mp, "Item Price": item_price, "Total Purchase Value": total_pp_mp})

#new_mos_prof_df["Item Price"] = new_mos_prof_df["Item Price"].map("${:,.2f}".format)
#new_mos_prof_df["Total Purchase Value"] = new_mos_prof_df["Total Purchase Value"].map("${:,.2f}".format)

#Format Item Price and Total Purchase Value--Sort by Total Purchase Value in descending order
new_mos_prof_df.head()
new_mos_prof_df.sort_values(["Total Purchase Value"], ascending=False)


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
...                                                              ...   
125     Whistling Mithril Warblade                                 2   
126     Exiled Mithril Longsword                                   1   
23      Crucifer                                                   1   
104     Gladiator's Glaive                                         1   
42      The Decapitator                                            1   

                                                      Item Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
82      Nirvana                                             4.90   
145     Fiery Glass Crusader                                4.58   
92      Final Critic                                        4.88   
103     Singed Scalpel                                      4.35   
...                                                          ...   
125     Whistling Mithril Warblade                          1.00   
126     Exiled Mithril Longsword                            2.00   
23      Crucifer                                            1.99   
104     Gladiator's Glaive                                  1.93   
42      The Decapitator                                     1.75   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
92      Final Critic                                                 39.04  
103     Singed Scalpel                                               34.80  
...                                                                    ...  
125     Whistling Mithril Warblade                                    2.00  
126     Exiled Mithril Longsword                                      2.00  
23      Crucifer                                                      1.99  
104     Gladiator's Glaive                                            1.93  
42      The Decapitator                                               1.75  

[183 rows x 3 columns]

Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
92      Final Critic                                              8   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

Observable Trends in Data

-Males account for the most total purchases and most total purchase dollars of optional items to improve playing experience
-There is nearly six times more males than females making purchases and is the largest group making purchases 
-Nearly 45% of the total purchase count was made by age group 20-24
-Oathbreaker, Last Hope of the Breaking storm had the most purchases and was the most profitable item

Other notes: There could be an opportunity to understand why females and Other/Non-disclosed groups are not making more purchases. Also, there is a largest percentage of purchases made from age groups 15-19 and 20-24. Is there an opportunity to appeal to more age groups north of 24? Why is Oathbreaker, Last Hope of the Breaking Storm purchased the most? Is there something about this option that can be used to enhance other items that are not as profitable?
